pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,264 @@
+	// Clock exported for testing.
+	Clock clock.Clock
+	// KubeletReadinessToggles contains an entry for every toggle between kubelet Ready->NotReady or NotReady->Ready state.
+	KubeletReadinessToggles []time.Time
+	// NodeReady indicates if the node is ready. Exported for testing.
+	NodeReady bool
+
+	client client.Client
+	// firstFailure stores the time of the first failed kubelet health check
+	firstFailure *time.Time
+	dbus         dbus.DBus
+	recorder     record.EventRecorder
+	// lastInternalIP stores the node internalIP
+	lastInternalIP netip.Addr
+	// getAddresses is a func which returns a slice of ip addresses on this node
+	getAddresses          func() ([]net.Addr, error)
+	kubeletHealthEndpoint string
+}
+
+// NewKubeletHealthChecker create a instance of a kubelet health check.
+func NewKubeletHealthChecker(client client.Client, clock clock.Clock, dbus dbus.DBus, recorder record.EventRecorder, getAddresses func() ([]net.Addr, error)) *KubeletHealthChecker {
+	return &KubeletHealthChecker{
+		client:                  client,
+		dbus:                    dbus,
+		Clock:                   clock,
+		recorder:                recorder,
+		getAddresses:            getAddresses,
+		KubeletReadinessToggles: []time.Time{},
+		kubeletHealthEndpoint:   defaultKubeletHealthEndpoint,
+	}
+}
+
+// Name returns the name of this health check.
+func (*KubeletHealthChecker) Name() string {
+	return "kubelet"
+}
+
+// HasLastInternalIP returns true if the node.InternalIP was stored.
+// Exported for testing.
+func (k *KubeletHealthChecker) HasLastInternalIP() bool {
+	return k.lastInternalIP.IsValid()
+}
+
+// SetKubeletHealthEndpoint set the kubeletHealthEndpoint.
+// Exported for testing.
+func (k *KubeletHealthChecker) SetKubeletHealthEndpoint(kubeletHealthEndpoint string) {
+	k.kubeletHealthEndpoint = kubeletHealthEndpoint
+}
+
+// Check performs the actual health check for the kubelet.
+func (k *KubeletHealthChecker) Check(ctx context.Context, node *corev1.Node) error {
+	log := logf.FromContext(ctx).WithName("kubelet")
+
+	// This mimics the old behavior of the kubelet-health-monitor which checks for
+	// to many NotReady->Ready toggles in a certain time and reboots the node in such a case.
+	if isNodeReady(node) && !k.NodeReady {
+		needsReboot := k.ToggleKubeletState()
+		log.Info("Kubelet became Ready", "readinessChanges", len(k.KubeletReadinessToggles), "timespan", toggleTimeSpan)
+		if needsReboot {
+			log.Info("Kubelet toggled between NotReady and Ready too often. Rebooting the node now")
+			k.recorder.Eventf(node, corev1.EventTypeWarning, "kubelet", "Kubelet toggled between NotReady and Ready at least %d times in a %s time window. Rebooting the node now", maxToggles, toggleTimeSpan)
+			err := k.dbus.Reboot()
+			if err != nil {
+				k.RevertToggleKubeletState()
+				k.recorder.Event(node, corev1.EventTypeWarning, "kubelet", "Rebooting the node failed")
+				return fmt.Errorf("rebooting the node failed %w", err)
+			}
+		}
+	}
```

Comment:
Would it make sense to extract this logic into a separate function similar to `ensureNodeInternalIP(...)` called below? It would help in keeping `Check(...)` concise and readable.

WDYT?